library/data/recipe/src/main/java/org/quiltmc/qsl/recipe/api/VanillaRecipeBuilders.java

```diff
@@ -0,0 +1,407 @@
+		if (cookTime < 0) throw new IllegalArgumentException("Cook time must be equal or greater than 0");
+
+		return new SmeltingRecipe(id, group, input, output, experience, cookTime);
+	}
+
+	/**
+	 * Returns a new blasting recipe.
+	 *
+	 * @param id         the identifier of the recipe
+	 * @param group      the group of the recipe
+	 * @param input      the input ingredient
+	 * @param output     the output item stack
+	 * @param experience the experience given
+	 * @param cookTime   the cook time in ticks
+	 * @return the blasting recipe
+	 */
+	public static BlastingRecipe blastingRecipe(Identifier id, String group, Ingredient input, ItemStack output, float experience, int cookTime) {
+		if (input == Ingredient.EMPTY) throw new IllegalArgumentException("Input cannot be empty.");
+		if (cookTime < 0) throw new IllegalArgumentException("Cook time must be equal or greater than 0");
+
+		return new BlastingRecipe(id, group, input, output, experience, cookTime);
+	}
+
+	/**
+	 * Returns a new smoking recipe.
+	 *
+	 * @param id         the identifier of the recipe
+	 * @param group      the group of the recipe
+	 * @param input      the input ingredient
+	 * @param output     the output item stack
+	 * @param experience the experience given
+	 * @param cookTime   the cook time in ticks
+	 * @return the smoking recipe
+	 */
+	public static SmokingRecipe smokingRecipe(Identifier id, String group, Ingredient input, ItemStack output, float experience, int cookTime) {
+		if (input == Ingredient.EMPTY) throw new IllegalArgumentException("Input cannot be empty.");
+		if (cookTime < 0) throw new IllegalArgumentException("Cook time must be equal or greater than 0");
+
+		return new SmokingRecipe(id, group, input, output, experience, cookTime);
+	}
+
+	/**
+	 * Returns a new campfire cooking recipe.
+	 *
+	 * @param id         the identifier of the recipe
+	 * @param group      the group of the recipe
+	 * @param input      the input ingredient
+	 * @param output     the output item stack
+	 * @param experience the experience given
+	 * @param cookTime   the cook time in ticks
+	 * @return the campfire cooking recipe
+	 */
+	public static CampfireCookingRecipe campfireCookingRecipe(Identifier id, String group, Ingredient input,
+	                                                          ItemStack output, float experience, int cookTime) {
+		if (input == Ingredient.EMPTY) throw new IllegalArgumentException("Input cannot be empty.");
+		if (cookTime < 0) throw new IllegalArgumentException("Cook time must be equal or greater than 0");
+
+		return new CampfireCookingRecipe(id, group, input, output, experience, cookTime);
+	}
+
+	/**
+	 * Represents the basis of a recipe builder.
+	 *
+	 * @param <Self>   the type of the recipe builder
+	 * @param <Result> the type of the recipe
+	 */
+	public static abstract class Builder<Self extends Builder<Self, Result>, Result extends Recipe<?>> {
```

Comment:
```suggestion
	public static abstract class Builder<SELF extends Builder<SELF, RESULT>, Result extends Recipe<?>> {
```

Reply:
A reason for this?
Is it just to avoid possible confusion with actual types?

Reply:
mostly yeah. It was slightly hard to understand at first